printf "0x%016zXULL" doesn't work

I stumbled upon this issue where \`tests/sanity_test_vectors.h' could not be properly generated - hashes in XSUM_XXH64_testdata[] where simply zeroed.

```
static const XSUM_testdata64_t XSUM_XXH64_testdata[] = {
    {    0, 0x0000000000000000ULL, 0x0000000000000000ULL }, /* XSUM_XXH64_testdata[0] */
    {    0, 0x000000009E3779B1ULL, 0x0000000000000000ULL }, /* XSUM_XXH64_testdata[1] */
    {    1, 0x0000000000000000ULL, 0x0000000000000000ULL }, /* XSUM_XXH64_testdata[2] */
    {    1, 0x000000009E3779B1ULL, 0x0000000000000000ULL }, /* XSUM_XXH64_testdata[3] */
(...)
static const XSUM_testdata64_t XSUM_XXH3_testdata[] = {
    {    0, 0x0000000000000000ULL, 0x0000000000000000ULL }, /* XSUM_XXH3_testdata[0] */
    {    0, 0x0000000085EBCA8DULL, 0x0000000000000000ULL }, /* XSUM_XXH3_testdata[1] */
    {    1, 0x0000000000000000ULL, 0x0000000000000000ULL }, /* XSUM_XXH3_testdata[2] */
    {    1, 0x0000000085EBCA8DULL, 0x0000000000000000ULL }, /* XSUM_XXH3_testdata[3] */
(...)
```

After changing `"0x%016zXULL"` to `"0x%016llx"` it turned out to be ok.

```
--- sanity_test_vectors_generator.c~	2024-03-08 00:44:54.000000000 +0000
+++ sanity_test_vectors_generator.c	2024-03-12 20:36:08.715699018 +0000
@@ -138,11 +138,11 @@ static void fprintf_XSUM_testdata32_t(FI
 }
 
 static void fprintf_XSUM_testdata64_t(FILE* fp, XSUM_testdata64_t const v) {
-    fprintf(fp, "{ %4d, 0x%016zXULL, 0x%016zXULL },", v.len, v.seed, v.Nresult);
+    fprintf(fp, "{ %4d, 0x%016llXULL, 0x%016llXULL },", v.len, v.seed, v.Nresult);
 }
 
 static void fprintf_XSUM_testdata128_t(FILE* fp, XSUM_testdata128_t const v) {
-    fprintf(fp, "{ %4d, 0x%016zXULL, { 0x%016zXULL, 0x%016zXULL } },",
+    fprintf(fp, "{ %4d, 0x%016llXULL, { 0x%016llXULL, 0x%016llXULL } },",
             v.len, v.seed, v.Nresult.low64, v.Nresult.high64);
 }
 
@@ -260,7 +260,7 @@ static void generate_sanity_test_vectors
 
     size_t index = 0;
     for(size_t len = 0; len < maxLen; ++len) {
-        static const size_t seeds[] = { 0, PRIME64 };
+        static const uint64_t seeds[] = { 0, PRIME64 };
         for(size_t iSeed = 0; iSeed < sizeof(seeds)/sizeof(seeds[0]); ++iSeed) {
             size_t const seed = seeds[iSeed];
             XSUM_testdata64_t const v = tvgen_XXH3_64bits_withSeed(sanityBuffer, len, seed);
@@ -312,7 +312,7 @@ static void generate_sanity_test_vectors
 
     size_t index = 0;
     for(size_t len = 0; len < maxLen; ++len) {
-        static const size_t seeds[] = { 0, PRIME32, PRIME64 };
+        static const uint64_t seeds[] = { 0, PRIME32, PRIME64 };
         for(size_t iSeed = 0; iSeed < sizeof(seeds)/sizeof(seeds[0]); ++iSeed) {
             XSUM_U64 const seed = seeds[iSeed];
             XSUM_testdata128_t const v = tvgen_XXH3_128bits_withSeed(sanityBuffer, len, seed);

```

Also `seeds[]` should be defined as 64-bit wide or they become 32-bit on 32-bit system.

Hi @tansy,
Thank you for the report.

You're right.  These `size_t` and `%zX` should be fixed.

### `size_t seeds[]`
`size_t` for `seeds[]` must be replaced with `XSUM_U64`.

### `%zX`

Possible solutions are

- [`%llx`](https://github.com/Cyan4973/xxHash/blob/dev/tests/collisions/main.c#L856-L857)
- [`%08X%08X` combo](https://github.com/Cyan4973/xxHash/blob/dev/tests/sanity_test.c#L128-L129)
- C99 standard `PRIu64` macro

### Test

It seems recent version of gcc cross compiler for 32 bit target reports this issue properly.

Therefore we can add CI test for this issue as the following commands

```
# Setup
sudo apt-get update
sudo apt-get install gcc-multilib g++-multilib qemu-utils qemu-user-static
sudo apt-get install gcc-arm-linux-gnueabi qemu-system-arm

# Build and test for ARM 32 bit (NG for now)
XCC="arm-linux-gnueabi-gcc" XEMU="qemu-arm-static" CPPFLAGS="-DXXH_VECTOR=XXH_SCALAR" LDFLAGS="-static" CC=$XCC RUN_ENV=$XEMU make -C tests clean sanity_test_vectors.h test_sanity
```

<details>
<summary>Actual error log:</summary>

```
arm-linux-gnueabi-gcc -Wall -Wextra -Wundef -g -static sanity_test_vectors_generator.c -o sanity_test_vectors_generator
sanity_test_vectors_generator.c: In function ‘fprintf_XSUM_testdata64_t’:
sanity_test_vectors_generator.c:141:32: warning: format ‘%zX’ expects argument of type ‘size_t’, but argument 4 has type ‘XSUM_U64’ {aka ‘long long unsigned int’} [-Wformat=]
  141 |     fprintf(fp, "{ %4d, 0x%016zXULL, 0x%016zXULL },", v.len, v.seed, v.Nresult);
      |                           ~~~~~^                             ~~~~~~
      |                                |                              |
      |                                unsigned int                   XSUM_U64 {aka long long unsigned int}
      |                           %016llX
sanity_test_vectors_generator.c:141:45: warning: format ‘%zX’ expects argument of type ‘size_t’, but argument 5 has type ‘XSUM_U64’ {aka ‘long long unsigned int’} [-Wformat=]
  141 |     fprintf(fp, "{ %4d, 0x%016zXULL, 0x%016zXULL },", v.len, v.seed, v.Nresult);
      |                                        ~~~~~^                        ~~~~~~~~~
      |                                             |                         |
      |                                             unsigned int              XSUM_U64 {aka long long unsigned int}
      |                                        %016llX
sanity_test_vectors_generator.c: In function ‘fprintf_XSUM_testdata128_t’:
sanity_test_vectors_generator.c:145:32: warning: format ‘%zX’ expects argument of type ‘size_t’, but argument 4 has type ‘XSUM_U64’ {aka ‘long long unsigned int’} [-Wformat=]
  145 |     fprintf(fp, "{ %4d, 0x%016zXULL, { 0x%016zXULL, 0x%016zXULL } },",
      |                           ~~~~~^
      |                                |
      |                                unsigned int
      |                           %016llX
  146 |             v.len, v.seed, v.Nresult.low64, v.Nresult.high64);
      |                    ~~~~~~
      |                     |
      |                     XSUM_U64 {aka long long unsigned int}
sanity_test_vectors_generator.c:145:47: warning: format ‘%zX’ expects argument of type ‘size_t’, but argument 5 has type ‘XXH64_hash_t’ {aka ‘long long unsigned int’} [-Wformat=]
  145 |     fprintf(fp, "{ %4d, 0x%016zXULL, { 0x%016zXULL, 0x%016zXULL } },",
      |                                          ~~~~~^
      |                                               |
      |                                               unsigned int
      |                                          %016llX
  146 |             v.len, v.seed, v.Nresult.low64, v.Nresult.high64);
      |                            ~~~~~~~~~~~~~~~
      |                                     |
      |                                     XXH64_hash_t {aka long long unsigned int}
sanity_test_vectors_generator.c:145:60: warning: format ‘%zX’ expects argument of type ‘size_t’, but argument 6 has type ‘XXH64_hash_t’ {aka ‘long long unsigned int’} [-Wformat=]
  145 |     fprintf(fp, "{ %4d, 0x%016zXULL, { 0x%016zXULL, 0x%016zXULL } },",
      |                                                       ~~~~~^
      |                                                            |
      |                                                            unsigned int
      |                                                       %016llX
  146 |             v.len, v.seed, v.Nresult.low64, v.Nresult.high64);
      |                                             ~~~~~~~~~~~~~~~~
      |                                                      |
      |                                                      XXH64_hash_t {aka long long unsigned int}
sanity_test_vectors_generator.c: In function ‘generate_sanity_test_vectors_xxh3’:
sanity_test_vectors_generator.c:22:17: warning: conversion from ‘long long unsigned int’ to ‘unsigned int’ changes value from ‘11400714785074694797’ to ‘2246822541’ [-Woverflow]
   22 | #define PRIME64 11400714785074694797ULL
      |                 ^~~~~~~~~~~~~~~~~~~~~~~
sanity_test_vectors_generator.c:263:44: note: in expansion of macro ‘PRIME64’
  263 |         static const size_t seeds[] = { 0, PRIME64 };
      |                                            ^~~~~~~
sanity_test_vectors_generator.c: In function ‘generate_sanity_test_vectors_xxh128’:
sanity_test_vectors_generator.c:22:17: warning: conversion from ‘long long unsigned int’ to ‘unsigned int’ changes value from ‘11400714785074694797’ to ‘2246822541’ [-Woverflow]
   22 | #define PRIME64 11400714785074694797ULL
      |                 ^~~~~~~~~~~~~~~~~~~~~~~
sanity_test_vectors_generator.c:315:53: note: in expansion of macro ‘PRIME64’
  315 |         static const size_t seeds[] = { 0, PRIME32, PRIME64 };
      |                                                     ^~~~~~~
```

</details>

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -443,6 +443,8 @@ jobs:
       run: |
         CPPFLAGS="-DXXH_VECTOR=XXH_SCALAR" LDFLAGS="-static" CC=$XCC RUN_ENV=$XEMU make clean check
         CPPFLAGS="-DXXH_VECTOR=XXH_NEON" CFLAGS="-O3 -march=armv7-a -fPIC -mfloat-abi=softfp -mfpu=neon-vfpv4" LDFLAGS="-static" CC=$XCC RUN_ENV=$XEMU make clean check
+        make -C tests clean
+        CPPFLAGS="-DXXH_VECTOR=XXH_SCALAR" LDFLAGS="-static" CC=$XCC RUN_ENV=$XEMU make -C tests sanity_test_vectors.h test_sanity
 
     - name: AARCH64 (XXH_VECTOR=[ scalar, NEON, SVE ])
       if: ${{ startsWith(matrix.name, 'AARCH64') }}

--- a/tests/sanity_test_vectors_generator.c
+++ b/tests/sanity_test_vectors_generator.c
@@ -138,11 +138,11 @@ static void fprintf_XSUM_testdata32_t(FILE* fp, XSUM_testdata32_t const v) {
 }
 
 static void fprintf_XSUM_testdata64_t(FILE* fp, XSUM_testdata64_t const v) {
-    fprintf(fp, "{ %4d, 0x%016zXULL, 0x%016zXULL },", v.len, v.seed, v.Nresult);
+    fprintf(fp, "{ %4d, 0x%016llXULL, 0x%016llXULL },", v.len, v.seed, v.Nresult);
 }
 
 static void fprintf_XSUM_testdata128_t(FILE* fp, XSUM_testdata128_t const v) {
-    fprintf(fp, "{ %4d, 0x%016zXULL, { 0x%016zXULL, 0x%016zXULL } },",
+    fprintf(fp, "{ %4d, 0x%016llXULL, { 0x%016llXULL, 0x%016llXULL } },",
             v.len, v.seed, v.Nresult.low64, v.Nresult.high64);
 }
 
@@ -208,7 +208,7 @@ static void generate_sanity_test_vectors_xxh32(FILE* fp, size_t maxLen) {
 
     size_t index = 0;
     for(size_t len = 0; len < maxLen; ++len) {
-        static const size_t seeds[] = { 0, PRIME32 };
+        static const uint64_t seeds[] = { 0, PRIME32 };
         for(size_t iSeed = 0; iSeed < sizeof(seeds)/sizeof(seeds[0]); ++iSeed) {
             size_t const seed = seeds[iSeed];
             XSUM_testdata32_t const v = tvgen_XXH32(sanityBuffer, len, seed);
@@ -234,7 +234,7 @@ static void generate_sanity_test_vectors_xxh64(FILE* fp, size_t maxLen) {
 
     size_t index = 0;
     for(size_t len = 0; len < maxLen; ++len) {
-        static const size_t seeds[] = { 0, PRIME32 };
+        static const uint64_t seeds[] = { 0, PRIME32 };
         for(size_t iSeed = 0; iSeed < sizeof(seeds)/sizeof(seeds[0]); ++iSeed) {
             size_t const seed = seeds[iSeed];
             XSUM_testdata64_t const v = tvgen_XXH64(sanityBuffer, len, seed);
@@ -260,7 +260,7 @@ static void generate_sanity_test_vectors_xxh3(FILE* fp, size_t maxLen) {
 
     size_t index = 0;
     for(size_t len = 0; len < maxLen; ++len) {
-        static const size_t seeds[] = { 0, PRIME64 };
+        static const uint64_t seeds[] = { 0, PRIME64 };
         for(size_t iSeed = 0; iSeed < sizeof(seeds)/sizeof(seeds[0]); ++iSeed) {
             size_t const seed = seeds[iSeed];
             XSUM_testdata64_t const v = tvgen_XXH3_64bits_withSeed(sanityBuffer, len, seed);
@@ -312,7 +312,7 @@ static void generate_sanity_test_vectors_xxh128(FILE* fp, size_t maxLen) {
 
     size_t index = 0;
     for(size_t len = 0; len < maxLen; ++len) {
-        static const size_t seeds[] = { 0, PRIME32, PRIME64 };
+        static const uint64_t seeds[] = { 0, PRIME32, PRIME64 };
         for(size_t iSeed = 0; iSeed < sizeof(seeds)/sizeof(seeds[0]); ++iSeed) {
             XSUM_U64 const seed = seeds[iSeed];
             XSUM_testdata128_t const v = tvgen_XXH3_128bits_withSeed(sanityBuffer, len, seed);

